x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/SqlQueryResponse.java

```diff
@@ -150,15 +156,35 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
                 }
                 builder.endArray();
             }
-            builder.startArray("rows");
-            for (List<Object> row : rows()) {
-                builder.startArray();
-                for (Object value : row) {
-                    value(builder, mode, value);
+            
+            if (columnar) {
+                // columns can be specified (for the first REST request for example), or not (on a paginated/cursor based request)
+                // if the columns are missing, we take the first rows' size as the number of columns
+                long columnsCount = columns != null ? columns.size() : 0;
```

Comment:
Imho this is superfluous and can be removed.
If the `rows().get(0).size() != columns.size()` (when columns are there -> 1st request) then maybe we should throw an exception as this is a bug.